genie-web/src/main/java/com/netflix/genie/web/aspect/HealthCheckMetricsAspect.java

```diff
@@ -0,0 +1,248 @@
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *     Unless required by applicable law or agreed to in writing, software
+ *     distributed under the License is distributed on an "AS IS" BASIS,
+ *     WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *     See the License for the specific language governing permissions and
+ *     limitations under the License.
+ *
+ */
+package com.netflix.genie.web.aspect;
+
+import com.google.common.collect.Maps;
+import com.netflix.genie.core.util.MetricsConstants;
+import com.netflix.genie.core.util.MetricsUtils;
+import com.netflix.spectator.api.Id;
+import com.netflix.spectator.api.Registry;
+import lombok.extern.slf4j.Slf4j;
+import org.aspectj.lang.JoinPoint;
+import org.aspectj.lang.ProceedingJoinPoint;
+import org.aspectj.lang.annotation.Around;
+import org.aspectj.lang.annotation.Aspect;
+import org.aspectj.lang.annotation.Before;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.boot.actuate.health.Health;
+import org.springframework.boot.actuate.health.Status;
+import org.springframework.stereotype.Component;
+
+import javax.annotation.Nullable;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Aspect woven into Spring Boot 'Health' machinery to publish metrics such as time taken, errors and other signals
+ * useful for dashboards and alerting.
+ *
+ * @author mprimi
+ * @since 3.2.4
+ */
+@Aspect
+@Component
+@Slf4j
+public class HealthCheckMetricsAspect {
+
+    private static final String HEALTH_ENDPOINT_TIMER_NAME = "genie.health.endpoint.timer";
+    private static final String HEALTH_INDICATOR_TIMER_METRIC_NAME = "genie.health.indicator.timer";
+    private static final String HEALTH_INDICATOR_COUNTER_METRIC_NAME = "genie.health.indicator.counter";
+    private static final String HEALTH_FAILURES_COUNTER_METRIC_NAME = "genie.health.failure.counter";
+    private static final String HEALTH_INDICATOR_CLASS_TAG_NAME = "healthIndicatorClass";
+    private static final String HEALTH_INDICATOR_NAME_TAG_NAME = "healthIndicatorName";
+
+    private final Registry registry;
+    private final Id healthEndpointTimerId;
+    private final Id healthIndicatorTimerId;
+    private final Id healthIndicatorCounterId;
+    private final Id healthIndicatorFailureCounterId;
+
+    /**
+     * Autowired constructor.
+     * @param registry metrics registry
+     */
+    @Autowired
+    public HealthCheckMetricsAspect(final Registry registry) {
+        this.registry = registry;
+        this.healthEndpointTimerId = registry.createId(HEALTH_ENDPOINT_TIMER_NAME);
+        this.healthIndicatorTimerId = registry.createId(HEALTH_INDICATOR_TIMER_METRIC_NAME);
+        this.healthIndicatorCounterId = registry.createId(HEALTH_INDICATOR_COUNTER_METRIC_NAME);
+        this.healthIndicatorFailureCounterId = registry.createId(HEALTH_FAILURES_COUNTER_METRIC_NAME);
```

Comment:
Any reason not to just tag the timer with success or failure or similar?

Reply:
Your two questions are related.
The join point that is able to capture the indicator time to complete, and publish the timer metric, does not have access to the status reported by the healtcheck. So it can only publish how long it took to evaluate.

The other counter metric has access to status (UP/DOWN/OUT_OF_SERVICE), etc, but does not have access to timing (since the join point is aggregation of the final status, after all indicators have been evaluated already).
Because of this, I cannnot publish a single metric that has both timing, count and status.

I have to publish timing with a timer and separately status with a counter.

Reply:
So you want the timer as well so that we can alert on long running checks?

Reply:
Yes if HCs start timing out this should give us all the info to see at glance what is happening.